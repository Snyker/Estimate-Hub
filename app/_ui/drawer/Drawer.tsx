'use client'
import React, {PropsWithChildren, useRef} from "react";
import {DrawerHook} from "@/app/_hooks/drawer.hook";

const classBySide = (side: Side) => {
    switch (side) {
        case "bottom":
            return 'drawer-bottom';
        case "right":
            return 'drawer-right';
        case 'top':
            return 'drawer-top';
        default:
            return ''
    }
}

export const Drawer: React.FC<PropsWithChildren<DrawerProps>> = (
    {
        drawer,
        children,
        side = "left",
        className = '',
        stopClose = false,
        open = false,
        ...props
    }) => {

    return (
        <div>
            <input type="checkbox" id={drawer.drawerId} className="drawer-toggle" ref={drawer.inputDrawerRef} />
            <label className="overlay" htmlFor={stopClose ? '' : drawer.drawerId}></label>
            <div className={`drawer w-fit max-w-screen-lg ${classBySide(side)} ${className}`} {...props}>
                {children}
            </div>
        </div>
    );
};

type Side = 'left' | 'top' | 'right' | 'bottom';

export type DrawerProps = React.HTMLAttributes<HTMLDivElement> & {
    drawer: DrawerHook;
    side?: Side;
    stopClose?: boolean;
    open?: boolean;
}
