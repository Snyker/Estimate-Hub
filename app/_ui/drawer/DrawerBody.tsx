import React, {HTMLAttributes, PropsWithChildren} from "react";

export const DrawerBody: React.FC<PropsWithChildren<HTMLAttributes<HTMLDivElement>>> = ({ children, className = '', ...props }) => {
    return (
        <div className={`drawer-content flex flex-col h-full ${className}`} {...props}>
            {children}
        </div>
    );
};
