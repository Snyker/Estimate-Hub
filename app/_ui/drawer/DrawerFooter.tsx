import React, {HTMLAttributes, PropsWithChildren} from "react";

export const DrawerFooter: React.FC<PropsWithChildren<HTMLAttributes<HTMLDivElement>>> = ({ children, className = '', ...props }) => {
    return (
        <div className={`drawer-footer ${className}`} {...props}>
            {children}
        </div>
    );
};
