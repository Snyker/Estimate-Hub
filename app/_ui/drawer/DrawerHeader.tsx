import React, {HTMLAttributes, PropsWithChildren} from "react";

export const DrawerHeader: React.FC<PropsWithChildren<HTMLAttributes<HTMLDivElement>>> = ({ children, className = '', ...props }) => {
    return (
        <div className={`drawer-header ${className}`} {...props}>
            {children}
        </div>
    );
};
