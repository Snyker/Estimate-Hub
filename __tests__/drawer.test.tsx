import React from 'react';
import renderer, {act} from 'react-test-renderer';
import {Drawer, DrawerProps} from "@/app/_ui/drawer/Drawer";
import {DrawerHeader} from "@/app/_ui/drawer/DrawerHeader";
import {DrawerFooter} from "@/app/_ui/drawer/DrawerFooter";
import {DrawerBody} from "@/app/_ui/drawer/DrawerBody";

describe('Drawer Component', () => {
    const defaultProps: DrawerProps = {
        children: 'Drawer content',
    };

    describe('Drawer', () => {

        it('renders correctly with default props', () => {
            const tree = renderer.create(<Drawer {...defaultProps} />).toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('applies custom class', () => {
            const tree = renderer.create(<Drawer {...defaultProps} className="custom-class" />).toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('applies side class correctly', () => {
            const tree = renderer.create(<Drawer {...defaultProps} side="right" />).toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('does not close when stopClose is true', () => {
            const tree = renderer.create(<Drawer {...defaultProps} stopClose={true} />).toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('renders as open when open is true', () => {
            const tree = renderer.create(<Drawer {...defaultProps} open={true} />).toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('handles onClick event', () => {
            const onClickMock = jest.fn();
            const tree = renderer.create(<Drawer {...defaultProps} onClick={onClickMock} />);

            act(() => {
                tree.root.props.onClick();
            })


            expect(onClickMock).toHaveBeenCalled();
        });

        it('renders content based on children', () => {
            const tree = renderer.create(<Drawer {...defaultProps}><div>Custom content</div></Drawer>).toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('renders additional HTML attributes', () => {
            const tree = renderer.create(<Drawer {...defaultProps} data-testid="drawer" />).toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('handles className collision between props and side class', () => {
            const tree = renderer.create(<Drawer {...defaultProps} className="drawer-right" />).toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('renders correctly with multiple complex props', () => {
            const onClickMock = jest.fn();
            const tree = renderer.create(
                <Drawer {...defaultProps} side="bottom" className="custom-class" stopClose={true} open={true} onClick={onClickMock}>
                    <div>Custom content</div>
                </Drawer>
            );
            expect(tree.toJSON()).toMatchSnapshot();
            act(() => {
                tree.root.props.onClick();
            })

            expect(onClickMock).toHaveBeenCalled();
        });
    })

    describe('DrawerBody', () => {
        it('renders correctly with default props', () => {
            const tree = renderer.create(<DrawerBody {...defaultProps} />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('applies custom class', () => {
            const tree = renderer.create(<DrawerBody {...defaultProps} className="custom-class" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('handles onClick event', () => {
            const onClickMock = jest.fn();
            const tree = renderer.create(<DrawerBody {...defaultProps} onClick={onClickMock} />);
            act(() => {
                tree.root.props.onClick();
            });
            expect(onClickMock).toHaveBeenCalled();
        });

        it('renders content based on children', () => {
            const tree = renderer.create(<DrawerBody {...defaultProps}><div>Custom content</div></DrawerBody>);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders additional HTML attributes', () => {
            const tree = renderer.create(<DrawerBody {...defaultProps} data-testid="drawer-body" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders correctly with multiple complex props', () => {
            const tree = renderer.create(
                <DrawerBody {...defaultProps} className="custom-class" onClick={() => {}} data-testid="drawer-body">
                    <div>Custom content</div>
                </DrawerBody>
            );
            expect(tree.toJSON()).toMatchSnapshot();
        });
    });

    describe('DrawerFooter', () => {
        it('renders correctly with default props', () => {
            const tree = renderer.create(<DrawerFooter {...defaultProps} />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('applies custom class', () => {
            const tree = renderer.create(<DrawerFooter {...defaultProps} className="custom-class" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders additional HTML attributes', () => {
            const tree = renderer.create(<DrawerFooter {...defaultProps} data-testid="drawer-footer" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders correctly with multiple complex props', () => {
            const tree = renderer.create(
                <DrawerFooter {...defaultProps} className="custom-class" data-testid="drawer-footer">
                    <div>Custom content</div>
                </DrawerFooter>
            );
            expect(tree.toJSON()).toMatchSnapshot();
        });
    });

    describe('DrawerHeader', () => {
        it('renders correctly with default props', () => {
            const tree = renderer.create(<DrawerHeader {...defaultProps} />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('applies custom class', () => {
            const tree = renderer.create(<DrawerHeader {...defaultProps} className="custom-class" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders additional HTML attributes', () => {
            const tree = renderer.create(<DrawerHeader {...defaultProps} data-testid="drawer-header" />);
            expect(tree.toJSON()).toMatchSnapshot();
        });

        it('renders correctly with multiple complex props', () => {
            const tree = renderer.create(
                <DrawerHeader {...defaultProps} className="custom-class" data-testid="drawer-header">
                    <div>Custom content</div>
                </DrawerHeader>
            );
            expect(tree.toJSON()).toMatchSnapshot();
        });
    });
});
