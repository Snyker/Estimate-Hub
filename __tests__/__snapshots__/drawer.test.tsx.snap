// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer Component Drawer applies custom class 1`] = `
<div
  className="drawer  custom-class"
>
  Drawer content
</div>
`;

exports[`Drawer Component Drawer applies side class correctly 1`] = `
<div
  className="drawer drawer-right "
>
  Drawer content
</div>
`;

exports[`Drawer Component Drawer does not close when stopClose is true 1`] = `
<div
  className="drawer  "
>
  Drawer content
</div>
`;

exports[`Drawer Component Drawer handles className collision between props and side class 1`] = `
<div
  className="drawer  drawer-right"
>
  Drawer content
</div>
`;

exports[`Drawer Component Drawer renders additional HTML attributes 1`] = `
<div
  className="drawer  "
  data-testid="drawer"
>
  Drawer content
</div>
`;

exports[`Drawer Component Drawer renders as open when open is true 1`] = `
<div
  className="drawer  "
>
  Drawer content
</div>
`;

exports[`Drawer Component Drawer renders content based on children 1`] = `
<div
  className="drawer  "
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Drawer Component Drawer renders correctly with default props 1`] = `
<div
  className="drawer  "
>
  Drawer content
</div>
`;

exports[`Drawer Component Drawer renders correctly with multiple complex props 1`] = `
<div
  className="drawer drawer-bottom custom-class"
  onClick={[MockFunction]}
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Drawer Component DrawerBody applies custom class 1`] = `
<div
  className="drawer-content flex flex-col h-full custom-class"
>
  Drawer content
</div>
`;

exports[`Drawer Component DrawerBody renders additional HTML attributes 1`] = `
<div
  className="drawer-content flex flex-col h-full "
  data-testid="drawer-body"
>
  Drawer content
</div>
`;

exports[`Drawer Component DrawerBody renders content based on children 1`] = `
<div
  className="drawer-content flex flex-col h-full "
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Drawer Component DrawerBody renders correctly with default props 1`] = `
<div
  className="drawer-content flex flex-col h-full "
>
  Drawer content
</div>
`;

exports[`Drawer Component DrawerBody renders correctly with multiple complex props 1`] = `
<div
  className="drawer-content flex flex-col h-full custom-class"
  data-testid="drawer-body"
  onClick={[Function]}
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Drawer Component DrawerFooter applies custom class 1`] = `
<div
  className="drawer-footer custom-class"
>
  Drawer content
</div>
`;

exports[`Drawer Component DrawerFooter renders additional HTML attributes 1`] = `
<div
  className="drawer-footer "
  data-testid="drawer-footer"
>
  Drawer content
</div>
`;

exports[`Drawer Component DrawerFooter renders correctly with default props 1`] = `
<div
  className="drawer-footer "
>
  Drawer content
</div>
`;

exports[`Drawer Component DrawerFooter renders correctly with multiple complex props 1`] = `
<div
  className="drawer-footer custom-class"
  data-testid="drawer-footer"
>
  <div>
    Custom content
  </div>
</div>
`;

exports[`Drawer Component DrawerHeader applies custom class 1`] = `
<div
  className="drawer-header custom-class"
>
  Drawer content
</div>
`;

exports[`Drawer Component DrawerHeader renders additional HTML attributes 1`] = `
<div
  className="drawer-header "
  data-testid="drawer-header"
>
  Drawer content
</div>
`;

exports[`Drawer Component DrawerHeader renders correctly with default props 1`] = `
<div
  className="drawer-header "
>
  Drawer content
</div>
`;

exports[`Drawer Component DrawerHeader renders correctly with multiple complex props 1`] = `
<div
  className="drawer-header custom-class"
  data-testid="drawer-header"
>
  <div>
    Custom content
  </div>
</div>
`;
